### PR TITLE
[BE][doc] explain why float8 needs to be distributed aware

### PR DESCRIPTION
**Pull request by weifengpy**
explain why float8 needs to be distributed aware and how it's different to bfloat16

### CODE REVIEW

File: docs/float8.md [hunk 1]
```diff
@@ -16,3 +16,16 @@ CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh --float8.enabl
 For parallelisms, we support float8 all-gather for FSDP (optional) and for TP (by default for `Float8Linear`).
 
 For scaling strategy, we currently support tensor-wise scaling with dynamic scales, and are actively working on tensor-wise scaling with delayed scales. Row-wise scaling is under exploration.
+
+## Why Composing Float8 with `torch.distributed`
+As shown below, for float8 for matmul, `torch._scaled_mm` requires both float8 tensor and scales. Scales are calculated from `max(abs)` of a high precision tensor.
```
Comment (awgu) [inline docs/float8.md]:
```suggestion
As shown below, for float8 for matmul, `torch._scaled_mm` requires both float8 tensors and their scales. Scales are calculated from `max(abs)` of a high precision tensor.
```

File: docs/float8.md [hunk 2]
```diff
@@ -16,3 +16,16 @@ CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh --float8.enabl
...
+```
+# float32/bfloat16 matmul, `torch.mm(input, weight)`, does not require scales
+# float8 matmul requires scales to ensure values can
```
Comment (awgu) [inline docs/float8.md]: nit: the newline here seems a bit early

File: docs/float8.md [hunk 3]
```diff
@@ -16,3 +16,16 @@ CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh --float8.enabl
...
+# fit within the representable range
+torch._scaled_mm(input_fp8, weight_fp8, scale_a=scale_input, scale_b=scale_weight)
+```
+
+Without considering distributed, we cast input and weight into float8 inside forward before calling torch._scaled_mm.
```
Comment (awgu) [inline docs/float8.md]:
```suggestion
Without considering distributed, we cast input and weight into float8 inside forward before calling `torch._scaled_mm`.
```

File: docs/float8.md [hunk 4]
```diff
@@ -16,3 +16,16 @@ CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh --float8.enabl
...
+
+Considering FSDP, we can cast high precision weights (1/N on each rank) into float8, and perform float8 all-gather to save bandwidth. At the beginning of the forward, we already have the unsharded float8 weights. Similarly, we can do float8 all-gather in TP to save bandwidth. The overhead is communicating `max(abs)` across ranks but for llama herd models, we have net wins in training speed.
```
Comment (awgu) [inline docs/float8.md]: I think we can clarify this part.
1. For both FSDP and TP, if we perform float8 all-gather, then we must also perform some communication of scales. As written, only the TP part mentions this.
2. For TP, when would float8 all-gather not be a net win? What part of "llama herd models" matters?

**Review by kwen2501 — APPROVED** [review]
LGTM.

Curious:
- what's the detailed type name used here?
- do we support calling all-gather on that dtype at the PyTorch API surface? Or do we need to make two separate calls?

File: docs/float8.md (hunk 4 above)
Comment (weifengpy) [reply docs/float8.md]: > As written, only the TP part mentions this

good point. will mention the communication for both FSDP and TP

> For TP, when would float8 all-gather not be a net win? What part of "llama herd models" matters?

float8 for TP should be strictly better. The amax communication overhead is meant for FSDP. llama herd models is probably less relevant. I can rephrase it to be more accurate, eg, "float8 all gather and amax communication can be a net win over float32/bfloat16 allgather, depending on world size and messge size."

**Comment by weifengpy** [comment]
> LGTM.
> 
> Curious:
> 
> * what's the detailed type name used here?
> * do we support calling all-gather on that dtype at the PyTorch API surface? Or do we need to make two separate calls?

> what's the detailed type name used here?

for matmul in the forward, input @ weight, dtypes are `torch.float8_e4m3fn` for both input and weight
for matmul in the backward, grad_output @ input, grad_output dtype is `torch.float8_e5m2 `, input dtype is `torch.float8_e4m3fn`

**Comment by tianyu-l** [comment]
Thanks for the BE work! It's great to have it documented, although as someone who doesn't know the details of how Float8 works, I found this note not clear enough.

E.g. you mentioned that this note
> explain why float8 needs to be distributed aware and how it's different to bfloat16

After reading the note, I only vaguely understood that we need to communicate the `max(abs)` values across all ranks for scaling purposes -- I guess you are saying: without the max, we don't know where the float point is, am I correct? But I don't know why bfloat16 doesn't -- I sort of know, but you didn't explain what's the difference.

> Without considering distributed, we cast input and weight into float8 inside forward before calling torch._scaled_mm.

I didn't understand this sentence. Are you saying: If we don't do distributed training (and instead do single-device training), then ...? Does it apply to every distributed training setting, e.g. what if I only use PP, is it still very different from single-device?

Again, this is from someone who don't understand things well ... hope it can provide some layman perspective :)

**Review by tianyu-l — CHANGES_REQUESTED** [review]
please see my review above

**Comment by weifengpy** [comment]
> But I don't know why bfloat16 doesn't -- I sort of know, but you didn't explain what's the difference.

I can mention this if it helps: for float8, it's user's responsibility to calculate scale. For bfloat16/float32, scale is managed internally and is not visible to user. Because user manages scales, 

> Does it apply to every distributed training setting, e.g. what if I only use PP, is it still very different from single-device?

PP is very similar to single-device training. I can change description to "In single-device training, ..."
